ipa-core/src/protocol/ipa_prf/validation_protocol/validation.rs

```diff
@@ -0,0 +1,508 @@
+            Vec::<Fp61BitPrime>::with_capacity(hashes.len());
+        challenges_from_proof_of_left_prover.push(hash_to_field(
+            &other_hashes.proof_from_prover_left_hash[0],
+            &hashes.proof_from_prover_left_hash[0],
+            exclude_large,
+        ));
+        for (left, right) in other_hashes
+            .proof_from_prover_left_hash
+            .iter()
+            .zip(hashes.proof_from_prover_left_hash.iter())
+        {
+            challenges_from_proof_of_left_prover.push(hash_to_field(left, right, exclude_small));
+        }
+
+        // compute challenges_from_proof_of_right_prover
+        // the prover computes hash_to_field(left,right)
+        // this is the left verifier of the party to the right,
+        // i.e. it has the left share, other is the right share
+        let mut challenges_from_proof_of_right_prover =
+            Vec::<Fp61BitPrime>::with_capacity(hashes.len());
+        challenges_from_proof_of_right_prover.push(hash_to_field(
+            &hashes.proof_from_prover_right_hash[0],
+            &other_hashes.proof_from_prover_right_hash[0],
+            exclude_large,
+        ));
+        for (left, right) in hashes
+            .proof_from_prover_right_hash
+            .iter()
+            .zip(other_hashes.proof_from_prover_right_hash.iter())
+        {
+            challenges_from_proof_of_right_prover.push(hash_to_field(left, right, exclude_small));
+        }
+
+        (
+            challenges_from_proof_of_left_prover,
+            challenges_from_proof_of_right_prover,
+        )
+    }
+}
+
+struct ProofHash {
+    proof_from_prover_left_hash: Vec<Hash>,
+    proof_from_prover_right_hash: Vec<Hash>,
+}
+
+impl ProofHash {
+    fn len(&self) -> usize {
+        assert_eq!(
+            self.proof_from_prover_left_hash.len(),
+            self.proof_from_prover_right_hash.len()
+        );
+        self.proof_from_prover_left_hash.len()
+    }
+    fn generate_hash(batch_to_verify: &BatchToVerify) -> Self {
+        let length = batch_to_verify.proofs_from_left_prover.len();
+        let mut proof_from_prover_left_hash = Vec::<Hash>::with_capacity(length);
+        let mut proof_from_prover_right_hash = Vec::<Hash>::with_capacity(length);
+
+        // compute first hash
+        proof_from_prover_left_hash.push(compute_hash(
+            batch_to_verify.first_proof_from_left_prover.iter(),
+        ));
+        proof_from_prover_right_hash.push(compute_hash(
+            batch_to_verify.first_proof_from_right_prover.iter(),
+        ));
+
+        // compute other hashes
+        // we don't need to hash last proof
+        for i in 0..(length - 1) {
+            proof_from_prover_left_hash.push(compute_hash(
+                batch_to_verify.proofs_from_left_prover[i].iter(),
+            ));
+            proof_from_prover_right_hash.push(compute_hash(
+                batch_to_verify.proofs_from_right_prover[i].iter(),
+            ));
+        }
```

Comment:
This `ProofHash` class has a LOT of duplicate code. It does the same thing for the left and then again for the right. It makes me feel like it would be better to change this struct to only contain hashes for one side, and then just instantiate two of them.